### ISSUE TITLE
Settings UI list should be focusless

### ISSUE THREAD

**Issue by joaomoreno** [Microsoft/vscode#54039]
I find the current settings editor very weird because I can select "rows" within:

![screenshot from 2018-07-11 13-59-46](https://user-images.githubusercontent.com/22350/42569998-b01c654c-8512-11e8-94e5-5ecef5a50bdb.png)

![screenshot from 2018-07-11 13-57-35](https://user-images.githubusercontent.com/22350/42569895-618ae886-8512-11e8-9360-9626fbed0a97.png)

It is also an accessibility nightmare, because a blind user will lose time focusing that list and its elements when what he should be focusing are the actual controls, just like a form in a HTML page.

We're currently using the focus to "expand" the description. This adds to the confusion because by selecting a row, the list relays out and other items move down. What we should do instead is present a `Read more` link, which would do that expansion, for each element.

**Comment by roblourens** [comment]
This keeps coming so I will explore it, but I think it will add a hurdle to the normal browsing use case, by making users interact individually with every setting just to read its description. I think that reading a setting happens much more than editing it so we should optimize for that case.

Also, scrolling off the page will be interesting. The tree is virtualized and controls don't exist until they are visible. I guess I can add a fake tab target at the bottom of the list, and when it's focused, reveal the next row and focus it?

**Comment by joaomoreno** [comment]
Why are we using a virtualized tree/list for this anyway? It's not like there are 10k settings. Why not simple HTML?

**Comment by roblourens** [comment]
There might not be 10k but there are 1k (with my extensions) and that won't get any smaller. There are input controls and event listeners and it gets redrawn when you search. Seems worth it.

**Comment by roblourens** [comment]
Are you expecting that 'tab' would tab through every link and control of every setting? How would a screenreader user read the setting description? And if they end up in the middle of the list, how do they tab out besides tabbing 1000 times? I think it's better to tab through the main components of the page. When the list takes focus, it will act like any other list in vscode. But at that point the user has to tab or press enter to focus the control, and I don't know how they learn how to do that.

**Comment by joaomoreno** [comment]
I expect it to work like a simple HTML form: https://developer.mozilla.org/en-US/docs/Learn/HTML/Forms/How_to_structure_an_HTML_form

> Are you expecting that 'tab' would tab through every link and control of every setting?

Yeah, that's usually how it works.

> How would a screenreader user read the setting description?

Using `<label>` elements et al.

> And if they end up in the middle of the list, how do they tab out besides tabbing 1000 times?

Just like any other control which supports search/filter: press a keybinding to narrow down the search. <kbd>Ctrl F</kbd> to go back to the search input field.

**Comment by roblourens** [comment]
I don't know what screenreader users expect, but I think if the description is in a <label> for the control, it will read the whole paragraph of description text before reading the value, when often they probably want to hear the title then skip to the value which we can do if we separate focusing the setting from focusing the control (like we do now).

**Comment by joaomoreno** [comment]
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/forms/Basic_form_hints

Check the `Describing with ARIA` section.

**Comment by isidorn** [comment]
Agree with some of @joaomoreno points.
Can we look at Keybindings editor as something that already solved a similar problem? First tab focus out of the search and onto the list. Once you are in the list you can use arrow keys to move focus.

On a side note: if a focus is in the settings editor but not inside some input field I would expect typing to go into the Search. This way users could narrow down their search as they are moving focus between elements.

**Comment by roblourens** [comment]
Thanks for the feedback everyone -

@cleidigh 

> It takes too many keystrokes to get to and change settings currently

There is a hidden shortcut that helps. If focus is in the search bar, you can press the down arrow key to focus the settings list. But 'tab' still has to stop everywhere in between.

I think this is a pro for the current model though. If the user can use the arrow keys to focus a setting, that's fewer keypresses than tabbing through many links and controls to get to the setting.

> Besides navigation to an individual setting, there is also an issue with the settings keyboard controller intercepting up, Down, PageUp, PageDown keys which are normally used to navigate a select

I see this but only on Windows - maybe we can fix this in the custom dropdown control.

> There appear to be scenarios where a screen reader seems to be reading out practically the entire list including a million 'clickable' items

I've seen this too, I have some screenreader issues to clear up...

> Header Rows probably should not be focusable, either means more keystrokes and/or screenreader output

Agreed, if we keep the current model. I have another issue tracking this.

> A usability issue here is also when you have clickable/tab indexed items in the description. e.g. Files.Autosave I have to Tab three times to get to the drop-down. If the detailed description were opened optionally I could skip that.

Another solution would be to make the control the first tab stop within the setting row so that I can tab directly from the row to the control, then tab past it to the links if I want. I do like expanding the description automatically to make browsing settings as easy as possible.

> I think I have gotten into certain situations where nothing seems to have focus 

Hm, I'd like to know more about that if you can reproduce it consistently.

@zersiax 

> Making all fields individually tabable doesn't sound like a good idea. It's going to be tricky to determine just from tabbing what belongs to what, as you would need some way of noting what a checkbox you just tabbed to actually toggles.

When you tab to a control, it would read the setting title, value, and description.

> my first thought when you described the scenario was to make it into a table semantically

That's interesting. I think I would prefer a model which is close to the layout on the screen. But I will keep that in mind.

> As for your first idea, that interaction model is rather non-standard; I'd rather expect right arrow to take me to a setting's controls if I'm using up and down arrows to move between settings. Tab wouldn't be my first thought, as usually tab would leave the list rather than move among controls on any given row.

Thanks for the feedback. I could have the right arrow focus a setting's control! The enter key will do it currently. But, the left arrow would probably go into the control, and not send focus back to the setting.

@isidorn 

> Can we look at Keybindings editor as something that already solved a similar problem? First tab focus out of the search and onto the list.

We have essentially the same model except that we have additional tab targets inside each row.

> On a side note: if a focus is in the settings editor but not inside some input field I would expect typing to go into the Search. This way users could narrow down their search as they are moving focus between elements.

Just like the new outline view, we could do that

**Comment by misolori** [comment]
Here's a concept where tabbing will go through all of the input items first, and then using the left/right arrows will navigate to the links/context menu. The difficult part here is making the context menu work on checkboxes + descriptions (that's a separate discussion, though). Thoughts?

![gif](https://user-images.githubusercontent.com/35271042/43794148-47b648d0-9a32-11e8-8c04-bd9e850e9701.gif)

**Comment by roblourens** [comment]
I've checked in a new implementation that is only based on tabbing, let me know what you think.

**Comment by joaomoreno** [comment]
Very cool. I'd close this 👍
